### PR TITLE
#158565106 Save new asset category 

### PR DESCRIPTION
**Pull request by FelixWambiri**
**What does this PR do?**
This PR makes use of the API endpoint to create new asset categories.

**Description of Task to be completed?**
* Enable users to create new asset categories.
* Notify the user if any of the required fields are not properly filled.

**How should this be manually tested?**
* visit `http://activo-dev.andela.com:3000/settings/categories/new` to create asset categories.

**Any background context you want to provide?**

The API endpoint currently used to create the asset categories is the mock API as the staging API is currently unavailable for use at the moment.

**What are the relevant pivotal tracker stories?**
[#158565106](https://www.pivotaltracker.com/story/show/158565106)

**Screenshots (if appropriate)**
image

**Questions:**

xxx

### CODE REVIEW

File: src/pages/CreateCategoryPage/CreateCategoryPage.test.tsx [hunk 1]
```diff
@@ -8,9 +8,5 @@ import { mount } from 'enzyme';
 import CreateCategoryPage from './';
 
 describe('The CreateCategoryPage component', () => {
-  it('renders a form for creating a new category', () => {
-    mount(<CreateCategoryPage />);
-
-    expect(mount).toMatchSnapshot();
-  });
+  it('renders a form for creating a new category');
```
Comment (Billmike) [inline src/pages/CreateCategoryPage/CreateCategoryPage.test.tsx]: No test for this?

File: src/pages/CreateCategoryPage/index.tsx [hunk 2]
```diff
@@ -9,82 +12,216 @@ import Attribute from 'components/Attribute';
...
+    return errors;
+    
+  }
+  /**
+   * This method handles what happens when the add category button is clicked
+   * 
+   * @param {Event} event
+   * @returns {void}
+   */
+  onSubmit = (event) => {
```
Comment (Billmike) [inline src/pages/CreateCategoryPage/index.tsx]: No tests for this?

File: src/pages/CreateCategoryPage/index.tsx [hunk 3]
```diff
@@ -9,82 +12,216 @@ import Attribute from 'components/Attribute';
...
+  }
+
+  /**
+   *  Set the errors object fields with `this field is required` if empty
+   * 
+   * @memberof CreateCategoryPage
+   * @param {Object} category
+   * @returns {object}
+   */
+  validateData = (data)  => {
```
Comment (Billmike) [inline src/pages/CreateCategoryPage/index.tsx]: Did you write a test for this method?

File: src/pages/CreateCategoryPage/index.tsx [hunk 4]
```diff
@@ -9,82 +12,216 @@ import Attribute from 'components/Attribute';
...
-      </form>
-    </div>
-  </React.Fragment>
-));
+      </React.Fragment>
+    );
+  }
+}
+
+const mapStateToProps = state => ({
```
Comment (Billmike) [inline src/pages/CreateCategoryPage/index.tsx]: Write tests for mapStateToProps and mapDispatchToProps for this component

File: src/pages/CreateCategoryPage/index.tsx [hunk 5]
```diff
@@ -9,82 +12,216 @@ import Attribute from 'components/Attribute';
...
+    });
+  }
+
+  /**
+   * Check if the errors object is empty
+   * 
+   * @param {Object} error
+   * @returns {boolean} 
+   */ 
+  hasErrors = (errors) => {
```
Comment (Billmike) [inline src/pages/CreateCategoryPage/index.tsx]: A lot of uncovered lines on your branch. You need to write tests to handle every logic implemented in your task, and that includes testing your class methods.

File: src/components/Attribute/interfaces.d.ts [hunk 6]
```diff
@@ -2,6 +2,9 @@ export interface AttributeProps {
   id?: string;
   enableConfigOptions?: boolean;
   error?: string;
+  name?: string;
+  onChange?: (e: Object) => any;
```
Comment (missating) [inline src/components/Attribute/interfaces.d.ts]: I think using `event` makes this clearer than using `e`

File: src/pages/CreateCategoryPage/index.tsx [hunk 7]
```diff
@@ -9,82 +12,218 @@ import Attribute from 'components/Attribute';
...
-            <h4 className="form-fields__heading__attribute-name">Attribute name</h4>
-            <h4 className="form-fields__heading__input-type">Input type</h4>
-          </div>
-
-          <Attribute
-            placeholder="Location"
-            error="This field is required."
-          />
+// actions
+import { addNewAssetCategory } from '../../store/modules/categories/index';
```
Comment (missating) [inline src/pages/CreateCategoryPage/index.tsx]: We already setup alias in our `webpack config` to resolve files so you don't have to do `'../../store/modules/categories/index'` when importing.

Do this instead `'modules/categories'`

File: src/pages/CreateCategoryPage/index.tsx [hunk 8]
```diff
@@ -9,82 +12,218 @@ import Attribute from 'components/Attribute';
...
 
-          <Attribute 
-            placeholder="Asset tag"
-          />
+// interfaces
+import { Errors, NewCategoryProps } from '../../store/modules/categories/interfaces';
```
Comment (missating) [inline src/pages/CreateCategoryPage/index.tsx]: Change this too.

File: src/store/modules/categories/index.ts [hunk 9]
```diff
@@ -1,8 +1,51 @@
...
+  ADD_NEW_CATEGORY_SUCCESS,
+  GET_CATEGORIES_FAILURE,
+  GET_CATEGORIES_SUCCESS
+} from './types';
+
+// components
+import { displayToastMessage } from '../toasts/index';
+
+// helper functions
+import { authService } from '../../../utils/auth';
```
Comment (missating) [inline src/store/modules/categories/index.ts]: The alias thing applies to both line 26 and 29

File: src/store/modules/categories/index.ts [hunk 10]
```diff
@@ -1,8 +1,51 @@
...
+ */ 
+const token = authService.getToken();
+const instance = axios.create({
+  baseURL: 'https://d5d7c01b-6aa3-4101-affc-23cabec2d75d.mock.pstmn.io',
+  headers: {
+    Authorization: `Bearer ${token}`,
+    'x-mock-response-code': 201,
+    ContentType: 'application/json',
+  },
+});
```
Comment (missating) [inline src/store/modules/categories/index.ts]: Extract this `axios` instance into a helper function.

The baseURL should be in the env file too

File: src/store/modules/categories/index.ts [hunk 11]
```diff
@@ -21,19 +61,64 @@ export const getAssetCategoriesFailure = (): CategoriesActionFailure => {
...
+ * @param category 
+ * @returns {object} action type and payload
+ */ 
+// export const addNewAssetCategory = (category, redirect) => (dispatch) => {
+export const addNewAssetCategory = category => (dispatch) => {
+  return instance.post('/api/v1/asset-categories', category)
+  .then((response) => {
+    dispatch(addAssetCategorySuccess(response.data));
+    dispatch(displayToastMessage('Asset Category successfully created', 'success'));
+    history.push('/dashboard');
```
[the remark this answers is not in the document]
Comment (missating) [reply src/store/modules/categories/index.ts]: Yeah. I agree with you. @Billmike

**Comment by kevinejiro** [comment]
The pivotal tracker ID in your PR should be a link to the actual story on the PT board.

**Comment by missating** [comment]
@FelixWambiri Why do you have 55 files changed? I'm very much surprised. :tired_face:

File: config/webpack.dev.js [hunk 12]
```diff
@@ -4,7 +4,7 @@ const path = require('path');
 const {
   namedModulesPlugin,
   hotModuleReplacementPlugin
-} = require('./webpack.plugins');
+} = require('./webpack.plugins.js');
```
Comment (missating) [inline config/webpack.dev.js]: I don't think there is any need for this change, in our `webpack config` we are already resolving file extensions including the `.js`, you don't need to add it when importing again. 

Please revert this change.

File: nightwatchGlobalSetup.js [hunk 13]
```diff
@@ -1,6 +1,6 @@
 import chromedriver from "chromedriver";
 
-import app from './app';
+import app from 'app';
```
Comment (missating) [inline nightwatchGlobalSetup.js]: Where exactly would `app` be imported from? 

Please revert this too.

File: server.js [hunk 14]
```diff
@@ -1,3 +1,3 @@
-const app = require('./app')
+const app = require('app')
```
Comment (missating) [inline server.js]: The `app` comment applies here too

File: src/components/AddAssetForm/index.tsx [hunk 15]
```diff
@@ -11,7 +11,7 @@ import RadioButtons from 'components/RadioButtons';
 import SelectBox from 'components/SelectBox';
 
 // interfaces
-import { AddAssetFormProps } from './interfaces';
+import { AddAssetFormProps } from 'components/AddAssetForm/interfaces';
```
Comment (missating) [inline src/components/AddAssetForm/index.tsx]: This `interface` file belongs in the same folder as the `AddAssetForm` component. There absolutely no need for you to use `components/AddAssetForm/interfaces` since its the same directory. 

Please revert here too and anywhere else this may apply.

File: src/components/AddAssetPane/AddAssetPane.test.tsx [hunk 16]
```diff
@@ -6,7 +6,7 @@ import { mount, shallow } from 'enzyme';
 import * as moment from 'moment';
 
 // components
-import AddAssetPane from './';
+import AddAssetPane from 'components/AddAssetPane';
```
Comment (missating) [inline src/components/AddAssetPane/AddAssetPane.test.tsx]: The former is better than your change.

This applies to other test files too

**Comment by FelixWambiri** [comment]
@missating I changed the way I had done importation of files in one component and this automatically updated in other components but I have reverted the changes in the other files.
